bootstrapper/internal/joinclient/joinclient.go

```diff
@@ -93,146 +90,134 @@ func New(lock locker, dial grpcDialer, joiner ClusterJoiner, meta MetadataAPI, l
-	c.cleaner = cleaner
+	for {
+		err := c.getNodeMetadata()
+		if err == nil {
+			c.log.With(slog.String("role", c.role.String()), slog.String("name", c.nodeName)).Info("Received own instance metadata")
+			break
+		}
+		c.log.With(slog.Any("error", err)).Error("Failed to retrieve instance metadata")
 
-	ticker := c.clock.NewTicker(c.interval)
-	go func() {
-		defer ticker.Stop()
-		defer func() { c.stopDone <- struct{}{} }()
-		defer c.log.Info("Client stopped")
-
-		diskUUID, err := c.getDiskUUID()
-		if err != nil {
-			c.log.With(slog.Any("error", err)).Error("Failed to get disk UUID")
-			return
+		c.log.With(slog.Duration("interval", c.interval)).Info("Sleeping")
+		select {
+		case <-c.stopC:
+			return nil
+		case <-ticker.C():
 		}
-		c.diskUUID = diskUUID
-
-		for {
-			err := c.getNodeMetadata()
-			if err == nil {
-				c.log.With(slog.String("role", c.role.String()), slog.String("name", c.nodeName)).Info("Received own instance metadata")
-				break
-			}
-			c.log.With(slog.Any("error", err)).Error("Failed to retrieve instance metadata")
-
-			c.log.With(slog.Duration("interval", c.interval)).Info("Sleeping")
-			select {
-			case <-c.stopC:
-				return
-			case <-ticker.C():
-			}
+	}
+
+	var ticket *joinproto.IssueJoinTicketResponse
+	var kubeletKey []byte
+
+	for {
+		ticket, kubeletKey, err = c.tryJoinWithAvailableServices()
+		if err == nil {
+			c.log.Info("Successfully retrieved join ticket, starting Kubernetes node")
+			break
 		}
+		c.log.With(slog.Any("error", err)).Warn("Join failed for all available endpoints")
 
-		for {
-			err := c.tryJoinWithAvailableServices()
-			if err == nil {
-				c.log.Info("Joined successfully. Client is shutting down")
-				return
-			} else if isUnrecoverable(err) {
-				c.log.With(slog.Any("error", err)).Error("Unrecoverable error occurred")
-				// TODO(burgerdev): this should eventually lead to a full node reset
-				return
-			}
-			c.log.With(slog.Any("error", err)).Warn("Join failed for all available endpoints")
-
-			c.log.With(slog.Duration("interval", c.interval)).Info("Sleeping")
-			select {
-			case <-c.stopC:
-				return
-			case <-ticker.C():
-			}
+		c.log.With(slog.Duration("interval", c.interval)).Info("Sleeping")
+		select {
+		case <-c.stopC:
+			return nil
+		case <-ticker.C():
 		}
-	}()
+	}
+
+	if err := c.startNodeAndJoin(ticket, kubeletKey, cleaner); err != nil {
+		c.log.With(slog.Any("error", err)).Error("Failed to start node and join cluster") // unrecoverable error
+		resetErr := c.markDiskForReset()
+		return errors.Join(err, resetErr)
+	}
+
+	return nil
 }
 
 // Stop stops the client and blocks until the client's routine is stopped.
 func (c *JoinClient) Stop() {
-	c.mux.Lock()
-	defer c.mux.Unlock()
-
-	if c.stopC == nil { // daemon not running
-		return
-	}
-
 	c.log.Info("Stopping")
 
 	c.stopC <- struct{}{}
 	<-c.stopDone
 
-	c.stopC = nil
-	c.stopDone = nil
+	c.stopC = make(chan struct{}, 1)
+	c.stopDone = make(chan struct{}, 1)
```

Comment:
I removed all code that was explicitly there to handle starting multiple `JoinClient.Start()` routines, since this was not (and should not be) done.
Resetting the channels here is mostly being done so calling `JoinClient.Start()` after `JoinClient.Stop()` does not panic.
This is because `c.stopC` and `c.stopDone` channels are no longer created during `JoinClient.Start()`, but during `joinclient.New()`, so setting them to `nil` during `JoinClient.Stop()` causes errors if `JoinClient.Start` is called after.

I think simply removing this code section would be functionally the same, and better highlights that we don't intend to reuse the client.